# Part 1. Equation of a Slime

How many late days are you using for this assignment?

In [2]:
# Imports section
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, KFold
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier

## 1. Loading the dataset

In [3]:
# Using pandas load the dataset
df = pd.read_csv('science_data_large.csv')
# Output the first 15 rows of the data
print(df.head(15))
# Display a summary of the table information (number of datapoints, etc.)
print(df.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## 2. Splitting the dataset

In [4]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = df[['Temperature °C', 'Mols KCL']] 
Y = df['Size nm^3']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.1, random_state=42)

## 3. Perform a Linear Regression

In [5]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, Y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
data_test = pd.DataFrame([['500','500']])
test_prediction = model.predict(data_test)
print(test_prediction)
# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
model_score = model.score(X_test, Y_test)
print(model_score)
# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
coef = model.coef_
intercept = model.intercept_
print('Coefficients -->', coef)
print('Intercept -->', intercept)

[540029.26034545]
0.8552472077276095
Coefficients --> [ 866.14641337 1032.69506649]
Intercept --> -409391.47958340764


c:\Users\kesha\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


Write the linear equation of a slime: (example equation: $E = mc^2$)

$$
h(x) = 866.146 \cdot Temperature + 1032.695 \cdot Mols - 409391.4796
$$

Report on score and explain meaning:
score of 0.860 means that 86% of the variation in Size nm^3 is explained by Temperature °C and Mols KCL, meaning the model captures most of the important patterns in the data. Since this is a high score, it indicates the model makes accurate predictions and should perform well on new data.

## 4. Use Cross Validation

In [6]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
cv = KFold(n_splits=5, shuffle=True, random_state=42)
cv_scores = cross_val_score(model, X, Y, cv=cv, scoring='r2')
print('Cross-validation scores:', cv_scores)
# Report on their finding and their significance

Cross-validation scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]


Write findings here: The cross-validation scores show that the model makes good predictions, as they range from 82.7% to 88.2%. Since the scores are similar across different data splits, it means the model is reliable and consistent, and will work well on new data

## 5. Using Polynomial Regression

In [7]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)
poly = PolynomialFeatures(degree=2)
X_train_p = poly.fit_transform(X_train)
X_test_p = poly.fit_transform(X_test)

model_poly = LinearRegression()
model_poly.fit(X_train_p, Y_train)

model_p_score = model_poly.score(X_test_p, Y_test)
print('Initial Score: ', model_p_score)

cv = KFold(n_splits=5, shuffle=True, random_state=42)
cv_scores_poly = cross_val_score(model_poly, X_train_p, Y_train, cv=cv)
print('Cross-Val Scores: ', cv_scores_poly)

coef = model_poly.coef_
intercept = model_poly.intercept_
print('Coefficients -->', coef)
print('Intercept -->', intercept)
# Report on the metrics and output the resultant equation as you did in Part 3.

Initial Score:  1.0
Cross-Val Scores:  [1. 1. 1. 1. 1.]
Coefficients --> [ 0.00000000e+00  1.20000000e+01 -1.27196606e-07  1.26450728e-11
  2.00000000e+00  2.85714287e-02]
Intercept --> 2.0478793885558844e-05


Write the polynomial equation of a slime: (example equation: $E = mc^2$)
$$
\small
\text{Size nm}^3 = (12.0000 \cdot \small\text{Temp}) + (-1.27196 \times 10^{-7} \cdot \small\text{Mols}) 
\\ + (1.2645 \times 10^{-11} \cdot \small\text{Temp}^2) + (2.0000 \cdot \small\text{Mols}^2) 
\\ + (0.0286 \cdot \text{Temp} \cdot \small\text{Mols}) + 2.0478 \times 10^{-5}
$$


Report on the score and interpret: A model score of 1.0 and cross-validation scores of 1.0 mean that the polynomial regression model perfectly predicts the target variable for both the test set and all cross-validation folds. This suggests overfitting, where the model has memorized the training data instead of learning general patterns, making it likely to perform poorly on new, unseen data.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [8]:
# Load the dataset. Then train and evaluate the classification models.
data = pd.read_csv('ckd_feature_subset.csv')

X_2 = data.drop(columns=["Target_ckd"])  # Features
Y_2 = data["Target_ckd"]  # Target variable

# Standardize features
scaler = StandardScaler()
X_2_scaled = scaler.fit_transform(X_2)

# Define cross-validation strategy
cv = KFold(n_splits=5, shuffle=True, random_state=42)

# Define classification models
class_models = {
    "Logistic Regression": LogisticRegression(random_state=42),
    "SVM (SVC)": SVC(kernel="rbf", random_state=42),
    "k-NN": KNeighborsClassifier(n_neighbors=5),
}

results = {}

for model_name, model in class_models.items():
    scores = cross_val_score(model, X_2_scaled, Y_2, cv=cv)
    results[model_name] = {"Mean Accuracy": np.mean(scores), "Std Dev": np.std(scores)}

print(results)

{'Logistic Regression': {'Mean Accuracy': 0.9479569892473119, 'Std Dev': 0.043569489112761005}, 'SVM (SVC)': {'Mean Accuracy': 0.9670967741935484, 'Std Dev': 0.02934203509816868}, 'k-NN': {'Mean Accuracy': 0.921505376344086, 'Std Dev': 0.060429724867728095}}


## Results and Conclusion for Classification Experiments
The SVM model performed the best, achieving the highest mean accuracy (96.7%) with a low standard deviation (2.9%), indicating both strong performance and stability. Logistic Regression followed closely with 94.8% accuracy, while k-NN had the lowest accuracy (92.2%) and the highest variability (6.0%), suggesting it may be more sensitive to data distribution. Overall, SVM is the best choice for this dataset due to its ability to capture complex patterns while maintaining consistent performance across different data splits.

In [9]:
neural_network_model = MLPClassifier(hidden_layer_sizes=(50,), max_iter=500, random_state=42)

nn_scores = cross_val_score(neural_network_model, X_2_scaled, Y_2, cv=cv, scoring="accuracy")

nn_results = {
    "Mean Accuracy": np.mean(nn_scores),
    "Std Dev": np.std(nn_scores)
}

print(nn_results)

c:\Users\kesha\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\kesha\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\kesha\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(


{'Mean Accuracy': 0.9608602150537635, 'Std Dev': 0.031522803356622414}


c:\Users\kesha\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(


## Results and Conclusion for Neural Network Experiments
The Neural Network achieved a mean accuracy of 96.1%, showing strong predictive performance. The standard deviation of 3.2% indicates consistent results across different data splits, meaning the model generalizes well. Compared to the Classification models, the performance of the Neural Network is only slightly worse than the SVC (best Classification).